Resume the IIS state machine on the thread pool

davidfowl opened the pull request:
- We post completion to the thread pool instead of the IIS thread pool to avoid having
new requests compete with existing threads for IIS threads.
We're going to use the local queue so that we can resume the state machine on this thread after
unwinding.

Fixes https://github.com/dotnet/aspnetcore/issues/41255

davidfowl committed:
Resume the IIS state machine on the thread pool
- We post completion to the thread pool instead of the IIS thread pool to avoid having
new requests compete with existing threads for IIS threads.
We're going to use the local queue so that we can resume the state machine on this thread after
unwinding.

BrennanConroy commented on src/Servers/IIS/IIS/src/Core/IISHttpContext.cs:
```diff
@@ -615,10 +615,24 @@ protected void ReportApplicationError(Exception ex)
+        },
+        // We use the native handle because we're going to dispose the safe handle by the time this runs.
+        // We don't want to hold onto it since managed code is unwinding.
+        (_requestNativeHandle.DangerousGetHandle(), requestNotificationStatus), preferLocal: true);
```
So this is safe to use inside the callback because we await the VTS right after calling PostCompletion? No one else can clean up the handle beneath us? Or could a server shutdown clean up the handle while the work item is queued?

davidfowl replied:
> So this is safe to use inside the callback because we await the VTS right after calling PostCompletion?

The VTS only protects the safe handle. This is the native handle, the only thing that cleans this up is native code. So when IIS itself is cleaning up.

This comment made me look at the native code though, I need to change it to set `IndicateManagedRequestComplete` in indicate completion.

davidfowl replied:
OK this made me look around:
- The native code checks if we're done with the managed request. I made a change to the code to set that boolean before we run this logic so it's similar to before.
- We have a new native -> managed code dependency to complete requests. I think that means if there's a native module blocking a thread, it'll end up blocking a thread pool thread. This didn't happen before as we were posting completions back to IIS before resuming the state machine.
- I guess it also means when shutdown is ungraceful managed code might be forcefully torn down when there's a native code running on a managed threads so AV?

HaoK replied:
>  shutdown is ungraceful managed code might be forcefully torn down when there's a native code running on a managed threads

I have a few issues assigned to me about our Stopping events not reliably firing, is this possibly related?

davidfowl replied:
I don't know if we can do anything useful in this case besides wait a little bit. Also testing it would require writing a misbehaving native module and blocking in it, then shutting down. I debugged the codez last night and have a little better understanding of the flow.

halter73 replied:
> I guess it also means when shutdown is ungraceful managed code might be forcefully torn down when there's a native code running on a managed threads so AV? 

I think what we want to do is to call `_requestNativeHandle.Dispose()` and await `await new ValueTask<object?>(_requestNativeHandle, _requestNativeHandle.Version);` before calling `CompleteHttpRequest()`. It was previously `http_post_completion` and now presumably `http_indicate_completion` that tells IIS that it's okay to release the backing handle.

The reason we have the weird dispose the SafeHandle then await it logic is to ensure any concurrent p/invokes with the SafeHandle have completed and any future ones will ODE. I think the logic was wrong here previously. We should have been doing this before calling `PostCompletion()` too.

Fortunately, this really only affects buggy code that tries to use the `HttpContext` after middleware complets. Still, it's better to avoid the possible AV.

davidfowl replied:
> I think what we want to do is to call _requestNativeHandle.Dispose() and await await new ValueTask<object?>(_requestNativeHandle, _requestNativeHandle.Version); before calling CompleteHttpRequest(). It was previously PostCompletion() and now presumably IndicateCompletion() that tells IIS that it's okay to release the backing handle.

The safe handle doesnt do much and isn't useful here. Indicate completion literally runs the rest of the request on the thread that calls it.

halter73 replied:
We should probably just update `NativeSafeHandle` to call `NativeMethods.HttpIndicateCompletion(handle, status)` and drop the `IValueTaskSource` stuff. You'd have to add a property or something for the status.

Then we could just dispatch the disposal of the `NativeSafeHandle`.

JamesNK commented on src/Servers/IIS/IIS/src/Core/IISHttpContext.cs:
```diff
@@ -615,10 +615,24 @@ protected void ReportApplicationError(Exception ex)
+        // unwinding.
+        ThreadPool.UnsafeQueueUserWorkItem(static state =>
+        {
+            var (handle, status) = state;
```
nit: There doesn't seem to be any benefit in making the callback static and passing in state vs a closure. The tuple will allocate when it's boxed to `object state`.

Not saying it should change, just curious whether there is a benefit I don't see, or habit of passing in state

davidfowl replied:
> nit: There doesn't seem to be any benefit in making the callback static and passing in state vs a closure. The tuple will allocate when it's boxed to object state.

This uses the generic overload https://docs.microsoft.com/en-us/dotnet/api/system.threading.threadpool.unsafequeueuserworkitem?view=net-6.0#system-threading-threadpool-unsafequeueuserworkitem-1(system-action((-0))-0-system-boolean). No boxing 😄

davidfowl committed:
Mark request complete before unwind.
- Add a new export to mark the managed request and call it after the request is over before resuming.

davidfowl committed:
Clean up tests looking for freb logs
- Remove async start and completion from freb logs.
- Dispose the client so tests run faster locally

HaoK commented on src/Servers/IIS/IIS/src/Core/IISHttpContext.cs:
```diff
@@ -615,10 +615,27 @@ protected void ReportApplicationError(Exception ex)
     }
 
-    public void PostCompletion(NativeMethods.REQUEST_NOTIFICATION_STATUS requestNotificationStatus)
+    // This method should only be called from the end of request processing, this lets use avoid most
```
nit
```suggestion
    // This method should only be called from the end of request processing, this lets us avoid most
```

HaoK commented on src/Servers/IIS/IIS/test/Common.FunctionalTests/FrebTests.cs:
```diff
@@ -56,6 +54,9 @@ public async Task CheckCommonFrebEvents()
         await result.HttpClient.GetAsync("HelloWorld");
 
+        // Dispose the client to speed up tests
+        result.HttpClient.Dispose();
```
Should we be doing this in all of our similar tests?  I can do that in a follow-up PR if this is noticeable

davidfowl commented on src/Servers/IIS/IIS/test/Common.FunctionalTests/FrebTests.cs:
```diff
@@ -107,6 +97,8 @@ public async Task CheckFrebDisconnect()
             await result.HttpClient.RetryRequestAsync("/WaitingRequestCount", async message => await message.Content.ReadAsStringAsync() == "1");
         }
 
+        result.HttpClient.Dispose();
```
We should probably do this will all of the tests, it would speed up running locally. These tests were taking 1 minute each now it's 3 seconds.

HaoK replied:
Hrm I'll try to make the update to the tests in a way where it will be easy to remove it (optimistically when in some future time this regression is fixed)

HaoK replied:
Is there some kind of keep alive on the http client side, something like https://makolyte.com/csharp-configuring-how-long-an-httpclient-connection-will-stay-open/#In_NET_Core

davidfowl replied:
It seems possible to disable connection pooling by setting the lifetime to TimeSpan.Zero. I'll try it.

davidfowl replied:
Works but I'll let you figure out how to apply it. I don't know if we should be changing the deployers directly or just these tests.

https://github.com/dotnet/aspnetcore/blob/4cdf42e78b7f56fc28ff4090a84674334e3694c2/src/Servers/IIS/IntegrationTesting.IIS/src/IISDeploymentResult.cs#L29

```
HttpClient = CreateClient(new SocketsHttpHandler() { PooledConnectionLifetime = TimeSpan.Zero });
```

halter73 commented:
Do we have any benchmarks showing how this affects RPS in a scenario like /json or /plaintext where there isn't a lot of latency?

davidfowl committed:
Move where we complete the request.

davidfowl commented:
Hmm this change broke it all 😄

HaoK commented:
I thought I heard you say you wanted to make things hang :)

davidfowl commented:
I see how it broke, I wrote some bad codez 😄

davidfowl committed:
Move all of the code into the safe handle

StephenBonikowsky commented:
@davidfowl is this PR being scrapped or still being worked on? Wondering what the plan is to fix #41255

davidfowl commented:
Not being scrapped, but put on hold. This helped but kicking up the load there were still differences. I haven't had time to spend on this for the last week or so, but I plan to get back to it soon.

japui commented:
Retome a máquina de estado do IIS no pool de threads #

davidfowl committed:
Merge branch 'main' into davidfowl/indicate-completion

BrennanConroy commented on src/Servers/IIS/IIS/src/Core/IISHttpContextSafeHandle.cs:
```diff
@@ -0,0 +1,52 @@
+    protected override bool ReleaseHandle()
+    {
+        // Dispose the context
+        Context!.Dispose();
```
If https://github.com/dotnet/aspnetcore/pull/46176 is merged, this Dispose call needs to run after `HttpIndicateCompletion`. Also, shouldn't `handle` below be disposed instead of relying on the finalizer?

davidfowl replied:
This isn't a finalizer

BrennanConroy replied:
The handle on line 40

amcasey commented:
> Not being scrapped, but put on hold. This helped but kicking up the load there were still differences. I haven't had time to spend on this for the last week or so, but I plan to get back to it soon.

@davidfowl Are you still planning to resuscitate this PR?

davidfowl commented:
> @davidfowl Are you still planning to resuscitate this PR?

Doesn't feel like it.

davidfowl commented:
Feels like we should attempt to resurrect this for 9.0 if we can verify the fix. My attempt made it better but I was never certain this fully solved the problem.
